lib/active_model/serializer.rb

```diff
@@ -102,24 +102,49 @@ def self.attributes(*attrs)
       end
     end
 
+    # TODO: remove the dynamic method definition
     # @example
     #   class AdminAuthorSerializer < ActiveModel::Serializer
     #     attributes :id, :recent_edits
     #     attribute :name, key: :title
     #
+    #     attribute :full_name do
+    #       "#{object.first_name} #{object.last_name}"
+    #     end
+    #
     #     def recent_edits
     #       object.edits.last(5)
-    #     enr
-    def self.attribute(attr, options = {})
+    #     end
+    def self.attribute(attr, options = {}, &block)
       key = options.fetch(:key, attr)
-      _attributes_keys[attr] = { key: key } if key != attr
-      _attributes << key unless _attributes.include?(key)
+      reader = if block
+                 ->(instance) { instance.instance_eval(&block) }
+               elsif _fragmented
+                 ->(instance) { instance.class._fragmented.read_attribute_for_serialization(attr) }
+               else
+                 ->(instance) { instance.read_attribute_for_serialization(attr) }
+               end
```

Comment:
Yeah, this starts getting rather ugly with if/elsif/else branches, a case seems like it could be a nice cleanup.

On the other hand, it does seem like pushing this switch based on the `_fragmented` status into `#read_attribute_for_serialization` would help keep that API clean, in that consumers who wanted to get an attribute wouldn't need to know anything about the fragment cache state.